exec/java-exec/src/test/java/org/apache/drill/test/ProfileParser.java

```diff
@@ -42,44 +48,207 @@
+          }
+          return result;
+        }
+      });
+      int currentFrag = 0;
+      int currentStep = 0;
+      for ( OpDefInfo opDef : raw ) {
+        if ( currentFrag < opDef.majorId ) {
+          currentFrag++;
+          OpDefInfo sender = new OpDefInfo( currentFrag, 0 );
+          sender.isInferred = true;
+          sender.name = "Sender";
+          sorted.add(sender);
+          currentStep = 1;
+          opDef.inferredParent = sender;
+          sender.children.add( opDef );
+        }
+        if ( opDef.stepId > currentStep ) {
+          OpDefInfo unknown = new OpDefInfo( currentFrag, currentStep );
+          unknown.isInferred = true;
+          unknown.name = "Unknown";
+          sorted.add(unknown);
+          opDef.inferredParent = unknown;
+          unknown.children.add( opDef );
+        }
+        sorted.add( opDef );
+        currentStep = opDef.stepId + 1;
+      }
+    }
+  }
+
+  /**
+   * Parse the plan portion of the query profile.
+   */
+
+  private void parsePlans() {
+    PlanParser parser = new PlanParser();
+    String plan = getPlan( );
+    parser.parsePlans(plan);
+    plans = parser.plans;
+    topoOrder = parser.operations;
+    operations = parser.sorted;
+  }
+
+  private void buildFrags() {
+    for (OpDefInfo opDef : operations) {
+      FragInfo major = fragments.get(opDef.majorId);
+      if (major == null) {
+        major = new FragInfo(opDef.majorId);
+        fragments.put(opDef.majorId, major);
+      }
+      major.ops.add(opDef);
+    }
+  }
+
+  private static List<FieldDef> parseCols(String cols) {
+    String parts[] = cols.split( ", " );
+    List<FieldDef> fields = new ArrayList<>( );
+    for ( String part : parts ) {
+      String halves[] = part.split( " " );
+      fields.add( new FieldDef( halves[1], halves[0] ) );
+    }
+    return fields;
+  }
+
+  private void parseFragProfiles() {
+    JsonArray frags = getFragmentProfile( );
+    for (JsonObject fragProfile : frags.getValuesAs(JsonObject.class)) {
+      int mId = fragProfile.getInt("majorFragmentId");
+      FragInfo major = fragments.get(mId);
+      major.parse(fragProfile);
+    }
+  }
+
+  private void mapOpProfiles() {
+    for (FragInfo major : fragments.values()) {
+      for (MinorFragInfo minor : major.minors) {
+        minor.mapOpProfiles(major);
+      }
+    }
+  }
+
+  /**
+   * A typical plan has many operator details across multiple
+   * minor fragments. Aggregate these totals to the "master"
+   * definition of each operator.
+   */
+
+  private void aggregateOpers() {
+    for (FragInfo major : fragments.values()) {
+      for (OpDefInfo opDef : major.ops) {
+        for ( OperatorProfile op : opDef.opExecs) {
+          Preconditions.checkState( major.id == op.majorFragId );
+          Preconditions.checkState( opDef.stepId == op.opId );
+          opDef.actualRows += op.records;
+          opDef.actualBatches += op.batches;
+          opDef.actualMemory += op.peakMem * 1024 * 1024;
```

Comment:
OK. Multiplications by constant powers of 2 is pretty fast so didn't worry about it originally.